### PR TITLE
De-duplicate unions.

### PR DESCRIPTION
**Pull request by mitchell-merry**
## Changes

If a union has multiple members that are the same, then we should de-duplicate them.

https://github.com/drwpow/openapi-typescript/pull/1049#discussion_r1159448414

## Checklist

- [x] Unit tests updated
- [ ] README updated
- [x] `examples/` directory updated (if applicable)

### CODE REVIEW

File: examples/github-api-next.ts [hunk 1]
```diff
@@ -91489,7 +91489,7 @@ export interface operations {
           /** @description The [status](https://docs.github.com/rest/commits/statuses) contexts to verify against commit status checks. If you omit this parameter, GitHub verifies all unique contexts before creating a deployment. To bypass checking entirely, pass an empty array. Defaults to all unique contexts. */
           required_contexts?: (string)[];
           payload?: OneOf<[{
-            [key: string]: unknown | undefined;
```
Comment (mitchell-merry) [inline examples/github-api-next.ts:91492]: this is pretty nice

**Comment by mitchell-merry** [comment]
Hm, I think this happens to fix https://github.com/drwpow/openapi-typescript/issues/810 as well.

Currently, a schema like this:
```yaml
    MetaQuote:
      oneOf: []
      discriminator:
        propertyName: meta_serialization_key
        mapping: {}
```

generates like so:
```ts
export interface components {
  schemas: {
    MetaQuote: ; // error here
  };
}
```

Now, we default to `never` when no types are passed into `tsUnionOf`, so we get:
```ts
export interface components {
  schemas: {
    MetaQuote: never;
  };
}
```

which I believe is what we expect, right?

File: src/utils.ts [hunk 2]
```diff
@@ -204,8 +204,41 @@ export function tsTupleOf(...types: string[]): string {
...
 export function tsUnionOf(...types: (string | number | boolean)[]): string {
-  if (types.length === 1) return String(types[0]); // don’t add parentheses around one thing
-  return types.map((t) => (TS_UNION_INTERSECTION_RE.test(String(t)) ? `(${t})` : t)).join(" | ");
+  if (types.length === 0) return "never";
+
+  // de-duplicate the union
+  const members = new Set<string>();
+  for (const t of types) {
+    // unknown swallows everything else in the union
+    if (t === "unknown") return "unknown";
```
Comment (drwpow) [inline src/utils.ts:213]: ah this is smart! I hadn’t considered this ❤️

File: src/utils.ts [hunk 3]
```diff
@@ -204,8 +204,41 @@ export function tsTupleOf(...types: string[]): string {
...
+
+    members.add(String(t));
+  }
+
+  // never gets swallowed by anything else, so only return never
+  //   if it is the only member
```
Comment (drwpow) [inline src/utils.ts:219]: 👍

**Review by drwpow — APPROVED** [review]

**Comment by drwpow** [comment]
Really appreciate all your awesome work on this library, and your helpfulness in the issues! I accept your offer to help and have invited you to have maintaner access 🙂. PRs still require approval from another maintainer, but you can at least open PRs and triage issues easier now. Thank you so much! 🙏

**Comment by drwpow** [comment]
When you get accept access to this project, I’ll let you merge this one 🙂

**Pull request merged**